neqo-transport/src/pmtud.rs

```diff
@@ -0,0 +1,644 @@
+            if self.probe_state == Probe::NotNeeded && now >= raise_timer {
+                qdebug!("PMTUD raise timer fired");
+                self.raise_timer = None;
+                self.start_pmtud();
+            }
+        }
+    }
+
+    /// Returns the current Packetization Layer Path MTU, i.e., the maximum UDP payload that can be
+    /// sent. During probing, this may be smaller than the actual path MTU.
+    #[must_use]
+    pub const fn plpmtu(&self) -> usize {
+        self.mtu - self.header_size
+    }
+
+    /// Returns true if a PMTUD probe should be sent.
+    #[must_use]
+    pub fn needs_probe(&self) -> bool {
+        self.probe_state == Probe::Needed
+    }
+
+    /// Returns the size of the current PMTUD probe.
+    #[must_use]
+    pub const fn probe_size(&self) -> usize {
+        self.search_table[self.probe_index] - self.header_size
+    }
+
+    /// Sends a PMTUD probe.
+    pub fn send_probe(&mut self, builder: &mut PacketBuilder, stats: &mut Stats) {
+        // The packet may include ACK-eliciting data already, but rather than check for that, it
+        // seems OK to burn one byte here to simply include a PING.
+        builder.encode_varint(FRAME_TYPE_PING);
+        stats.frame_tx.ping += 1;
+        stats.frame_tx.all += 1;
+        stats.pmtud_tx += 1;
+        self.probe_count += 1;
+        self.probe_state = Probe::Sent;
+        qdebug!(
+            "Sending PMTUD probe of size {}, count {}",
+            self.search_table[self.probe_index],
+            self.probe_count
+        );
+    }
+
+    /// Returns true if the packet is a PMTUD probe.
+    #[must_use]
+    pub fn is_pmtud_probe(&self, p: &SentPacket) -> bool {
+        self.probe_state == Probe::Sent && p.len() == self.probe_size()
+    }
+
+    /// Count the PMTUD probes included in `pkts`.
+    fn count_pmtud_probes(&self, pkts: &[SentPacket]) -> usize {
+        pkts.iter().filter(|p| self.is_pmtud_probe(p)).count()
+    }
+
+    /// Checks whether a PMTUD probe has been acknowledged, and if so, updates the PMTUD state.
+    /// May also initiate a new probe process for a larger MTU.
+    pub fn on_packets_acked(&mut self, acked_pkts: &[SentPacket], stats: &mut Stats) {
+        // Reset the loss counts for all packets sizes <= the size of the largest ACKed packet.
+        let max_len = acked_pkts.iter().map(SentPacket::len).max().unwrap_or(0);
+        if max_len == 0 {
+            // No packets were ACKed, nothing to do.
+            return;
+        }
+
+        let idx = self
+            .search_table
+            .iter()
+            .take_while(|&&sz| sz <= max_len + self.header_size)
+            .count();
+        self.loss_counts.iter_mut().take(idx).for_each(|c| *c = 0);
+
+        let acked = self.count_pmtud_probes(acked_pkts);
+        if acked == 0 {
+            return;
+        }
+
+        // A probe was ACKed, confirm the new MTU and try to probe upwards further.
+        stats.pmtud_ack += acked;
```

Comment:
Something for later perhaps: Maybe we should be tracking stats on a per-probe-size basis rather than just the total number of successful probes.